### ISSUE TITLE
Map GRPCAsyncResponseStream to built-in type?

### ISSUE THREAD

**Issue by eseay** [grpc/grpc-swift#1455]
### What are you trying to achieve?

I am currently using grpc-swift in an "internal SDK" of sorts that wraps all of the networking and much of the business logic for a certain area in my app. Though the host application transitively depends on grpc-swift, I've made an effort to abstract all the "gRPC stuff" away beneath my own APIs.

For unary calls, this has been seamless and is working great; however, now that I am attempting to implement a streaming call, I am curious the best way forward.

I'll also add that I am doing some custom decoding in our error handling code to inspect trailers for known objects that our server throws back, so I'd like for my wrapper to be able to do that as well.

### What have you tried so far?

My app is entirely built on Swift Concurrency, but this is somehow my first experience working with `AsyncSequence` and `AsyncStream`. I see that the stream function(s) return a `GRPCAsyncResponseStream`, which is obviously a grpc-swift type. I also see in the docs for that type that it is a `a wrapper around AsyncThrowingStream`. Is there a way I can map the `GRPCAsyncResponseStream` into a literal `AsyncThrowingStream` of my expected type so that the caller is unaware of the gRPC-based implementation details under the hood?

**Comment by glbrntt** [comment]
You can use this `init` from `AsyncThrowingStream`: https://github.com/apple/swift/blob/bc1443e43e59e6a693af73eb74b5633f7883fcc9/stdlib/public/Concurrency/AsyncThrowingStream.swift#L334-L388 which takes a function which asynchronously produces values. You can use the iterator from the `GRPCAsyncResponseStream` as that function. It would be something like:

```swift
var responseStream = ...
var responseStreamIterator = responseStream.makeAsyncIterator()
let stream = AsyncThrowingStream<Response, Error> { 
  try await responseStreamIterator.next()
}
```

**Comment by eseay** [comment]
> You can use this `init` from `AsyncThrowingStream`: https://github.com/apple/swift/blob/bc1443e43e59e6a693af73eb74b5633f7883fcc9/stdlib/public/Concurrency/AsyncThrowingStream.swift#L334-L388 which takes a function which asynchronously produces values. You can use the iterator from the `GRPCAsyncResponseStream` as that function. It would be something like:
> 
> ```swift
> var responseStream = ...
> var responseStreamIterator = responseStream.makeAsyncIterator()
> let stream = AsyncThrowingStream<Response, Error> { 
>   try await responseStreamIterator.next()
> }
> ```

🤝 Thank you! This looks like exactly what I need.

For my own curiosity and better understanding, would you mind providing high-level context on what is meant by the comment in the header below? Why not just return the `AsyncThrowingStream` as you demonstrate in the snippet above?

```swift
/// This is currently a wrapper around AsyncThrowingStream because we want to be
/// able to swap out the implementation for something else in the future.
@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
public struct GRPCAsyncResponseStream<Element: Sendable>: AsyncSequence {
```
